include/groov/mmio_bus.hpp

```diff
@@ -58,28 +106,46 @@ template <typename HardwareInterface = cpp_mem_iface> struct mmio_bus {
     static auto write(auto addr, decltype(Mask) value) -> async::sender auto {
         using base_type = decltype(Mask);
 
-        constexpr std::size_t lsb = std::countr_zero(Mask);
-        constexpr std::size_t width = std::countr_one(Mask >> lsb);
-        constexpr bool byte_aligned = (lsb % 8) == 0;
-        constexpr bool contiguous = std::popcount(Mask) == width;
-        constexpr bool pow2_width = std::has_single_bit(width);
-        constexpr bool accessible_width = width & (8u | 16u | 32u | 64u);
-        constexpr bool subword_access_en =
-            byte_aligned && contiguous && pow2_width && accessible_width;
-
-        if constexpr (subword_access_en) {
-            using subword_t = detail::uint_of_width_t<width>;
-            constexpr auto byte_offset = lsb / 8;
-            auto const subword_addr = addr + byte_offset;
-            auto const subword_val = static_cast<subword_t>(value >> lsb);
+        using subwords_aligned =
+            detail::mp_filter_q<detail::aligned_for<iface>,
+                                detail::subword_permuations>;
+
+        using subwords_aligned_and_fit =
+            detail::mp_filter_q<detail::fits_within<base_type>,
+                                subwords_aligned>;
+
+        using subword_candidate_index =
+            detail::mp_find_if_q<subwords_aligned_and_fit,
+                                 detail::subword_satisfies<Mask, IdMask>>;
+
+        constexpr bool found_subword =
+            subword_candidate_index::value <
+            detail::mp_size<subwords_aligned_and_fit>::value;
+
+        if constexpr (found_subword) {
+            using subword = detail::mp_at<subwords_aligned_and_fit,
+                                          subword_candidate_index>;
+            using subword_t = typename subword::subword_t;
+
+            constexpr auto id_write_mask =
+                subword::template mask<base_type> & ~Mask;
+            auto const write_value = value | (id_write_mask & IdValue);
```

Comment:
**issue**: This isn't right. We already know that the subword is covered by the value and the identity together, so how about:
```cpp
auto const write_value = (value | IdValue) & subword::template mask<base_type>;
```
?